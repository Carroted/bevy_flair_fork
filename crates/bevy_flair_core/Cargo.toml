[package]
name = "bevy_flair_core"
version = "0.2.0"
edition = "2021"
description = "Bevy UI styling using CSS"
license = "MIT OR Apache-2.0"
categories = ["games", "game-development"]
keywords = ["style", "animations", "bevy", "css"]
repository = "https://github.com/eckz/bevy_flair"
rust-version = "1.80.0"

[lints.rust]
unsafe_code = "allow"
missing_docs = "warn"

[dependencies]
thiserror = "2.0"
serde = "1.0"
rustc-hash = "2.1"

# Same version as bevy so it implements reflect
smol_str = "0.2"

bevy = { workspace = true }