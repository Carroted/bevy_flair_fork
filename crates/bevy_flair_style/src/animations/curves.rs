use bevy::math::cubic_splines::CubicSegment;
use bevy::math::curve::cores::*;
use bevy::math::curve::*;
use bevy::math::Vec2;
use bevy::reflect::Reflect;

/// A [`Curve`] whose sample space is timing is modified by a cubic bezier function.
#[derive(Copy, Clone, Debug, Reflect)]
pub struct CubicBezierEaseCurve {
    cubic_segment: CubicSegment<Vec2>,
}

impl Curve<f32> for CubicBezierEaseCurve {
    fn domain(&self) -> Interval {
        Interval::UNIT
    }

    fn sample_unchecked(&self, t: f32) -> f32 {
        self.cubic_segment.ease(t)
    }

    fn sample_clamped(&self, t: f32) -> f32 {
        self.cubic_segment.ease(t)
    }
}

impl CubicBezierEaseCurve {
    /// Create a new [`CubicBezierEaseCurve`] over the [unit interval] from a [`CubicSegment`].
    ///
    /// [unit interval]: `Interval::UNIT`
    pub fn new(p1: impl Into<Vec2>, p2: impl Into<Vec2>) -> Self {
        Self {
            cubic_segment: CubicSegment::new_bezier(p1, p2),
        }
    }

    /// Equal to [`Self::new`]`((0.25, 0.1), (0.25, 1.0))`, the default value, increases in velocity towards the middle of the transition, slowing back down at the end.
    pub fn ease() -> Self {
        Self::new((0.25, 0.1), (0.25, 1.))
    }

    /// Equal to [`Self::new`]`((0.42, 0), (1.0, 1.0))`, starts off slowly, with the transition speed increasing until complete.
    pub fn ease_in() -> Self {
        Self::new((0.42, 0.), (1., 1.))
    }

    /// Equal to [`Self::new`]`((0, 0), (0.58, 1.0))`, starts transitioning quickly, slowing down as the transition continues.
    pub fn ease_out() -> Self {
        Self::new((0., 0.), (0.58, 1.))
    }

    /// Equal to [`Self::new`]`((0.42, 0), (0.58, 1.0))`, starts transitioning slowly, speeds up, and then slows down again.
    pub fn ease_in_out() -> Self {
        Self::new((0.42, 0.), (0.58, 1.))
    }
}

/// A [`Curve`] that is defined by
///
/// - an initial `start` sample value at `t = 0`
/// - a final `end` sample value at `t = 1`
/// - an interpolation function to interpolate between the two values.
///
/// The resulting curve's domain is always [the unit interval].
///
/// [the unit interval]: Interval::UNIT
#[derive(Clone, Debug)]
pub struct LinearCurve<T, I> {
    /// Initial `start` sample value at `t = 0`.
    pub start: T,
    /// Final `end` sample value at `t = 1`.
    pub end: T,
    /// Interpolation function to interpolate between the two values.
    pub interpolation: I,
}

impl<T, I> Curve<T> for LinearCurve<T, I>
where
    I: Fn(&T, &T, f32) -> T,
{
    #[inline]
    fn domain(&self) -> Interval {
        Interval::UNIT
    }

    fn sample_unchecked(&self, t: f32) -> T {
        (self.interpolation)(&self.start, &self.end, t)
    }
}

/// A curve that is defined by explicit neighbor interpolation over a set of evenly-spaced samples.
/// Each sample has a defined easing function associated
pub struct SampleEasedCurve<T, I, E> {
    pub(crate) core: EvenCore<(T, E)>,
    pub(crate) interpolation: I,
}

#[allow(unused)]
impl<T, I, E> SampleEasedCurve<T, I, E> {
    /// Create a new [`SampleEasedCurve`] using the specified `interpolation` to interpolate between
    /// the given `samples`. An error is returned if there are not at least 2 samples or if the
    /// given `domain` is unbounded.
    ///
    /// The interpolation takes two values by reference together with a scalar parameter and
    /// produces an owned value. The expectation is that `interpolation(&x, &y, 0.0)` and
    /// `interpolation(&x, &y, 1.0)` are equivalent to `x` and `y` respectively.
    pub(crate) fn new(
        domain: Interval,
        samples: impl IntoIterator<Item = (T, E)>,
        interpolation: I,
    ) -> Result<Self, EvenCoreError>
    where
        I: Fn(&T, &T, f32) -> T,
        E: Curve<f32>,
    {
        Ok(Self {
            core: EvenCore::new(domain, samples)?,
            interpolation,
        })
    }
}

impl<T, I, E> Curve<T> for SampleEasedCurve<T, I, E>
where
    T: Clone,
    I: Fn(&T, &T, f32) -> T,
    E: Curve<f32>,
{
    #[inline]
    fn domain(&self) -> Interval {
        self.core.domain()
    }

    #[inline]
    fn sample_unchecked(&self, t: f32) -> T {
        self.sample_clamped(t)
    }

    #[inline]
    fn sample_clamped(&self, t: f32) -> T {
        // `EvenCore::sample_interp` is implicitly clamped.
        match self.core.sample_interp(t) {
            InterpolationDatum::Exact((sample, _))
            | InterpolationDatum::LeftTail((sample, _))
            | InterpolationDatum::RightTail((sample, _)) => sample.clone(),
            InterpolationDatum::Between((left_sample, easing), (right_sample, _), t) => {
                let eased_t = easing.sample_clamped(t);
                (self.interpolation)(left_sample, right_sample, eased_t)
            }
        }
    }
}

/// A curve that is defined by interpolation over unevenly spaced samples with explicit
/// interpolation.
/// Each sample has a defined easing function associated
///
/// A difference with [`UnevenSampleCurve`] is that this curve's domain is always [the unit interval].
/// [the unit interval]: Interval::UNIT
pub struct UnevenSampleEasedCurve<T, I, E> {
    pub(crate) core: UnevenCore<(T, E)>,
    pub(crate) interpolation: I,
}

impl<T, I, E> UnevenSampleEasedCurve<T, I, E> {
    /// Create a new [`UnevenSampleEasedCurve`] using the provided `interpolation` to interpolate
    /// between adjacent `timed_samples`. The given samples are filtered to finite times and
    /// sorted internally; if there are not at least 2 valid timed samples, an error will be
    /// returned.
    ///
    /// The interpolation takes two values by reference together with a scalar parameter and
    /// produces an owned value. The expectation is that `interpolation(&x, &y, 0.0)` and
    /// `interpolation(&x, &y, 1.0)` are equivalent to `x` and `y` respectively.
    ///
    /// The t value of the interpolation will be eased but the most-left sampled easing function
    pub fn new(
        timed_samples: impl IntoIterator<Item = (f32, (T, E))>,
        interpolation: I,
    ) -> Result<Self, UnevenCoreError>
    where
        I: Fn(&T, &T, f32) -> T,
        E: Curve<f32>,
    {
        Ok(Self {
            core: UnevenCore::new(timed_samples)?,
            interpolation,
        })
    }
}

impl<T, I, E> Curve<T> for UnevenSampleEasedCurve<T, I, E>
where
    T: Clone,
    I: Fn(&T, &T, f32) -> T,
    E: Curve<f32>,
{
    #[inline]
    fn domain(&self) -> Interval {
        Interval::UNIT
    }

    #[inline]
    fn sample_unchecked(&self, t: f32) -> T {
        match self.core.sample_interp(t) {
            InterpolationDatum::Exact((sample, _))
            | InterpolationDatum::LeftTail((sample, _))
            | InterpolationDatum::RightTail((sample, _)) => sample.clone(),
            InterpolationDatum::Between((left_sample, easing), (right_sample, _), t) => {
                let eased_t = easing.sample_unchecked(t);
                (self.interpolation)(left_sample, right_sample, eased_t)
            }
        }
    }
}
