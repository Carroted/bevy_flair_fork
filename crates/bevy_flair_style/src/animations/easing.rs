use bevy::math::{
    Vec2,
    curve::{
        Curve, Interval, UnevenSampleAutoCurve,
        easing::{EaseFunction as BevyEaseFunction, JumpAt},
    },
};
use bevy::reflect::Reflect;

use crate::animations::curves::CubicBezierEaseCurve;
use serde::{Deserialize, Serialize};

/// Represents the position of the step in a [`EasingFunction::Steps`].
///
/// This is equivalent to the CSS [`step-position`](https://developer.mozilla.org/en-US/docs/Web/CSS/easing-function/steps#step-position).
#[derive(Copy, Clone, Debug, PartialEq, Eq, Default, Reflect, Serialize, Deserialize)]
pub enum StepPosition {
    /// Indicates that the first step happens when the animation begins.
    JumpStart,
    /// Indicates that the last step happens when the animation ends.
    JumpEnd,
    /// Indicates neither early nor late jumps happen.
    JumpNone,
    /// Indicates both early and late jumps happen.
    JumpBoth,
    /// Indicates that the first step happens when the animation begins.
    Start,
    #[default]
    /// Indicates that the last step happens when the animation ends.
    End,
}

impl From<StepPosition> for JumpAt {
    fn from(value: StepPosition) -> Self {
        match value {
            StepPosition::JumpStart => JumpAt::Start,
            StepPosition::JumpEnd => JumpAt::End,
            StepPosition::JumpNone => JumpAt::None,
            StepPosition::JumpBoth => JumpAt::Both,
            StepPosition::Start => JumpAt::Start,
            StepPosition::End => JumpAt::End,
        }
    }
}

/// Represents [`EasingFunction`] as a [`Curve<f32>`].
#[derive(Clone, Debug, Reflect)]
pub(crate) enum EasingFunctionCurve {
    CubicBezier(CubicBezierEaseCurve),
    BevyEaseFunction(BevyEaseFunction),
    LinearPoints(UnevenSampleAutoCurve<f32>),
    SingleLinearPoint(f32),
    Linear,
}

impl Curve<f32> for EasingFunctionCurve {
    #[inline]
    fn domain(&self) -> Interval {
        Interval::UNIT
    }

    #[inline]
    fn sample_unchecked(&self, t: f32) -> f32 {
        self.sample_clamped(t)
    }

    #[inline]
    fn sample_clamped(&self, t: f32) -> f32 {
        match self {
            EasingFunctionCurve::CubicBezier(curve) => curve.sample_clamped(t),
            EasingFunctionCurve::BevyEaseFunction(curve) => curve.sample_clamped(t),
            EasingFunctionCurve::LinearPoints(curve) => curve.sample_clamped(t),
            EasingFunctionCurve::SingleLinearPoint(point) => *point,
            EasingFunctionCurve::Linear => t.clamp(0.0, 1.0),
        }
    }
}

/// Represents an easing function that can be used in animations.
///
/// It's the equivalent to the CSS [`<easing-function>`](https://developer.mozilla.org/en-US/docs/Web/CSS/easing-function).
#[derive(Clone, PartialEq, Debug, Default, Reflect, Serialize, Deserialize)]
pub enum EasingFunction {
    /// Transitions at an even speed.
    Linear,
    /// Creates a transition curve that progresses uniformly between points.
    LinearPoints(Vec<(f32, f32)>),
    /// The default value, increases in velocity towards the middle of the transition, slowing back down at the end.
    #[default]
    Ease,
    /// Starts off slowly, with the transition speed increasing until complete.
    EaseIn,
    /// Starts transitioning quickly, slowing down as the transition continues.
    EaseOut,
    /// Starts transitioning slowly, speeds up, and then slows down again.
    EaseInOut,
    /// An author-defined cubic-Bezier curve, where the p1 and p2 values must be in the range of 0 to 1
    CubicBezier {
        /// First control point of the cubic-bezier curve.
        p1: Vec2,
        /// Second control point of the cubic-bezier curve.
        p2: Vec2,
    },
    /// Creates stepped transitions that divides the animation into a set number of equal-length intervals.
    Steps {
        /// Number of steps
        steps: i32,
        /// Position of the step
        pos: StepPosition,
    },
}

impl EasingFunction {
    pub(crate) fn into_easing_curve(self) -> EasingFunctionCurve {
        match self {
            EasingFunction::Linear => EasingFunctionCurve::Linear,
            EasingFunction::LinearPoints(mut points) => {
                if points.is_empty() {
                    EasingFunctionCurve::Linear
                } else if points.len() == 1 {
                    EasingFunctionCurve::SingleLinearPoint(points[0].0)
                } else {
                    points.extend([(0.0, 0.0), (1.0, 1.0)]);
                    let curve = UnevenSampleAutoCurve::new(points).unwrap();

                    debug_assert_eq!(
                        curve.domain(),
                        Interval::UNIT,
                        "Linear curve has points outside of the [0.0, 1.0] range"
                    );
                    EasingFunctionCurve::LinearPoints(curve)
                }
            }
            EasingFunction::Ease => EasingFunctionCurve::CubicBezier(CubicBezierEaseCurve::ease()),
            EasingFunction::EaseIn => {
                EasingFunctionCurve::CubicBezier(CubicBezierEaseCurve::ease_in())
            }
            EasingFunction::EaseOut => {
                EasingFunctionCurve::CubicBezier(CubicBezierEaseCurve::ease_out())
            }
            EasingFunction::EaseInOut => {
                EasingFunctionCurve::CubicBezier(CubicBezierEaseCurve::ease_in_out())
            }
            EasingFunction::CubicBezier { p1, p2 } => {
                EasingFunctionCurve::CubicBezier(CubicBezierEaseCurve::new(p1, p2))
            }
            EasingFunction::Steps { steps, pos } => EasingFunctionCurve::BevyEaseFunction(
                BevyEaseFunction::Steps(steps as usize, pos.into()),
            ),
        }
    }
}
