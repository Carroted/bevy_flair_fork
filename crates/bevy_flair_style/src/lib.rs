//! # Bevy Flair Style
//! Bevy Flair Style is a styling system for Bevy UI. It allows you to style your UI using CSS-like syntax.
//!
//! This crate contains all the necessary components, systems and plugins to style your UI.
use bevy::prelude::*;
use bevy_flair_core::*;
use serde::{Deserialize, Serialize};

mod builder;
pub mod components;
mod simple_selector;
mod style_sheet;

pub mod animations;

#[cfg(test)]
mod testing;

#[cfg(feature = "css_selectors")]
pub mod css_selector;

mod systems;

use crate::animations::ReflectAnimationsPlugin;
use crate::components::*;

pub use builder::*;
pub use simple_selector::*;
pub use style_sheet::*;

pub(crate) type IdName = smol_str::SmolStr;
pub(crate) type ClassName = smol_str::SmolStr;
pub(crate) type TypeName = smol_str::SmolStr;

/// Represents the current state of an entity.
/// By default, it supports only the basic pseudo classes like `:hover`, `:active`, and `:focus`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default, Reflect, Serialize, Deserialize)]
pub struct NodeState {
    /// If the entity is pressed
    pub pressed: bool,
    /// If the entity is hovered
    pub hovered: bool,
    /// If the entity is focused
    pub focused: bool,
    /// If the entity is focused and the focus visibility is active
    pub focused_and_visible: bool,
}

impl NodeState {
    /// Empty StylePseudoState.
    pub const EMPTY: NodeState = NodeState {
        pressed: false,
        hovered: false,
        focused: false,
        focused_and_visible: false,
    };
    /// Returns true if state represents a pressed state.
    pub fn is_pressed(&self) -> bool {
        self.pressed
    }

    /// Checks if the state matches against a [`NodePseudoStateSelector`] .
    pub fn matches(&self, selector: NodePseudoStateSelector) -> bool {
        match selector {
            NodePseudoStateSelector::Pressed => self.pressed,
            NodePseudoStateSelector::Hovered => self.hovered,
            NodePseudoStateSelector::Focused => self.focused,
            NodePseudoStateSelector::FocusedAndVisible => self.focused_and_visible,
        }
    }
}

/// Helper class that can be used to match against a node state.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum NodePseudoStateSelector {
    /// If the entity is pressed
    Pressed,
    /// If the entity is hovered
    Hovered,
    /// If the entity is focused
    Focused,
    /// If the entity is focused and the focus indicator is visible
    FocusedAndVisible,
}

/// System sets for the [`FlairStylePlugin`] plugin.
#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug, SystemSet)]
pub enum StyleSystemSets {
    /// All changes that requires adding, or removing any component are applied here
    StructuralChanges,
    /// Apply changes to [`NodeStyleData`] component.
    SetStyleData,
    /// Mark all nodes that needs their style recalculated.
    MarkNodesForRecalculation,

    /// Moves transitions and animations forward.
    /// Happens before CalculateStyle, where new transitions and animations are added,
    /// so new animations are not moved on the frame they are added.
    TickAnimations,

    /// Calculates new properties, transitions and animations for nodes that are marked for recalculation
    CalculateStyle,

    /// Effectively applies changes from animations and style changes into the Components.
    ApplyProperties,
}

/// Plugin that adds the styling systems to Bevy.
pub struct FlairStylePlugin;

impl Plugin for FlairStylePlugin {
    fn build(&self, app: &mut App) {
        app.init_asset::<StyleSheet>()
            .register_type::<(
                NodeStyleSheet,
                NodeStyleData,
                NodeStyleMarker,
                ClassList,
                Siblings,
                NodeProperties,
            )>()
            .register_required_components::<Node, NodeStyleSheet>()
            .add_plugins((
                ReflectAnimationsPlugin,
                TrackTypeNameComponentPlugin::<Node>::new(0),
                TrackTypeNameComponentPlugin::<Label>::new(1),
                TrackTypeNameComponentPlugin::<Text>::new(1),
                TrackTypeNameComponentPlugin::<Button>::new(2),
            ))
            .configure_sets(
                PostUpdate,
                (
                    (
                        StyleSystemSets::StructuralChanges,
                        StyleSystemSets::SetStyleData,
                        StyleSystemSets::MarkNodesForRecalculation,
                        StyleSystemSets::CalculateStyle,
                        StyleSystemSets::ApplyProperties.before(bevy::ui::UiSystem::Prepare),
                    )
                        .chain(),
                    StyleSystemSets::TickAnimations.before(StyleSystemSets::CalculateStyle),
                ),
            )
            .add_systems(PreUpdate, systems::mark_as_changed_on_style_sheet_change)
            .add_systems(
                PostUpdate,
                (
                    (
                        // systems::add_styled_node_on_hierarchy_events,
                        // systems::on_root_added,
                        systems::calculate_effective_style_sheet,
                        systems::sync_siblings_system,
                    )
                        .chain()
                        .in_set(StyleSystemSets::StructuralChanges),
                    (
                        systems::calculate_is_root,
                        systems::apply_classes,
                        systems::interaction_system,
                        systems::track_name_changes,
                        systems::sync_input_focus,
                    )
                        .in_set(StyleSystemSets::SetStyleData),
                    systems::mark_nodes_for_recalculation
                        .in_set(StyleSystemSets::MarkNodesForRecalculation),
                    systems::tick_animations.in_set(StyleSystemSets::TickAnimations),
                    systems::calculate_style.in_set(StyleSystemSets::CalculateStyle),
                    systems::apply_properties.in_set(StyleSystemSets::ApplyProperties),
                ),
            );
    }
}

/// Tracks when a given component is present in the entity and adds the type name to the entity.
/// This is useful for matching against types like `Node`, or `Button`.
/// It's really difficult to know if an entity contains a type by name in Bevy,
/// And tracking all component would be unfeasible, so this plugin allows you to track only the components you need.
///
/// Each component is tracked with a priority, so if an entity has multiple components, the one with the highest priority will be used.
#[derive(Clone, Debug)]
pub struct TrackTypeNameComponentPlugin<T> {
    /// The priority of the component name.
    pub priority: u32,
    /// Optional override name of the component.
    pub name: Option<&'static str>,
    _marker: std::marker::PhantomData<fn() -> T>,
}

impl<C: Component + TypePath> TrackTypeNameComponentPlugin<C> {
    /// Creates a new instance of the plugin.
    pub fn new(priority: u32) -> Self {
        Self {
            priority,
            name: None,
            _marker: std::marker::PhantomData,
        }
    }

    #[allow(clippy::type_complexity)]
    fn on_added(
        &self,
    ) -> impl FnMut(Query<&mut NodeStyleData, Or<(Added<C>, (With<C>, Added<NodeStyleData>))>>)
    {
        let priority = self.priority;
        let name = self.name.unwrap_or(C::short_type_path());
        move |mut query| {
            for mut data in &mut query {
                data.push_type_name_with_priority(name, priority);
            }
        }
    }
}

impl<C: Component + TypePath> Plugin for TrackTypeNameComponentPlugin<C> {
    fn build(&self, app: &mut App) {
        app.add_systems(
            PostUpdate,
            self.on_added().in_set(StyleSystemSets::SetStyleData),
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bevy::app::App;
    use bevy::asset::weak_handle;
    use bevy::ecs::system::RunSystemOnce;
    use bevy::ui::Node;
    use std::sync::atomic;

    const TEST_STYLE_SHEET: NodeStyleSheet =
        NodeStyleSheet::new(weak_handle!("fe981062-17ce-46e4-999a-5a61ea8fe722"));

    const ROOT: (TestNode, NodeStyleSheet) = (TestNode, TEST_STYLE_SHEET);

    #[derive(Copy, Clone, Component)]
    #[require(Node)]
    struct TestNode;

    #[derive(Component)]
    struct FirstChild;

    #[derive(Copy, Clone, Component, Reflect)]
    #[reflect(Component)]
    #[require(Node)]
    struct Child;

    #[derive(Copy, Clone, Component, Reflect)]
    #[reflect(Component)]
    #[require(Node)]
    struct GrandChild;

    macro_rules! query_len {
        ($app:expr, $filter:ty) => {{
            let world = $app.world_mut();
            let mut query_state = world.query_filtered::<(), $filter>();
            query_state.iter(world).len()
        }};
    }

    macro_rules! query {
        ($app:expr, $components:ty) => {{
            let world: &mut World = $app.world_mut();
            let mut query_state = world.query::<$components>();
            query_state.iter(world).cloned().collect::<Vec<_>>()
        }};

        ($app:expr, $components:ty, $filter:ty) => {{
            let world: &mut World = $app.world_mut();
            let mut query_state = world.query_filtered::<$components, $filter>();
            query_state.iter(world).collect::<Vec<_>>()
        }};
    }

    macro_rules! assert_world_snapshot {
        ($app:expr, ( $($components:path),*) ) => {{
            let all_entities = query!($app, Entity, With<NodeStyleData>);
            let type_registry = $app.world().resource::<AppTypeRegistry>().0.read();

            let mut scene_builder = bevy::scene::DynamicSceneBuilder::from_world($app.world());
            scene_builder = scene_builder
                .allow_component::<Name>()
                .allow_component::<ChildOf>()
                .allow_component::<Children>()
                .allow_component::<NodeStyleSheet>()
                $(.allow_component::<$components>())*
                .extract_entities(all_entities.into_iter())
                .remove_empty_entities()
            ;

            let scene = scene_builder.build();
            let scene_serializer = bevy::scene::serde::SceneSerializer::new(&scene, &type_registry);

            insta::assert_ron_snapshot!(scene_serializer);
        }};
    }

    fn app() -> App {
        let mut app = App::new();

        app.init_resource::<PropertiesRegistry>();

        app.add_plugins((
            bevy::time::TimePlugin,
            AssetPlugin::default(),
            BevyUiPropertiesPlugin,
            FlairStylePlugin,
        ));

        app.register_type::<(Child, GrandChild)>();
        app.register_type_data::<std::collections::BinaryHeap<TypeNameWithPriority>, ReflectSerialize>();

        app.finish();

        app
    }

    #[test]
    fn spawn_many_children() {
        let mut app = app();

        app.add_systems(Startup, |mut commands: Commands| {
            commands
                .spawn((Name::new("Root"), ROOT))
                .with_children(|children| {
                    children.spawn(Child);
                    children.spawn(Child);
                    children.spawn(Child);
                });
        });

        app.update();

        assert_world_snapshot!(app, (Child, Siblings));
    }

    #[test]
    fn append_child_after_initial_spawn() {
        let mut app = app();

        let root_entity_id = app
            .world_mut()
            .run_system_once(|mut commands: Commands| {
                commands
                    .spawn((Name::new("Root"), ROOT))
                    .with_children(|children| {
                        children.spawn(Child);
                        children.spawn(Child);
                    })
                    .id()
            })
            .unwrap();

        app.update();

        app.world_mut()
            .run_system_once(move |mut commands: Commands| {
                commands.entity(root_entity_id).with_child(Child);
            })
            .unwrap();

        app.update();

        assert_world_snapshot!(app, (Child, Siblings, NodeStyleData));
    }

    #[test]
    fn spawn_many_children_and_grandchildren() {
        let mut app = app();

        app.add_systems(Startup, |mut commands: Commands| {
            commands
                .spawn((Name::new("Root"), ROOT))
                .with_children(|children| {
                    children.spawn(Child).with_children(|children| {
                        children.spawn(GrandChild);
                        children.spawn(GrandChild);
                        children.spawn(GrandChild);
                    });
                    children.spawn(Child);
                    children
                        .spawn((Child, NodeStyleSheet::Block))
                        .with_children(|children| {
                            children.spawn(GrandChild);
                            children.spawn(GrandChild);
                            children.spawn(GrandChild);
                        });
                });
        });

        app.update();
        assert_world_snapshot!(app, (Child, GrandChild, NodeStyleData, Siblings));
    }

    #[test]
    fn append_grand_children_after_initial_spawn() {
        let mut app = app();

        let root_entity_id = app
            .world_mut()
            .run_system_once(|mut commands: Commands| {
                commands
                    .spawn((Name::new("Root"), ROOT))
                    .with_children(|children| {
                        children.spawn(Child);
                        children.spawn((Child, NodeStyleSheet::Block));
                    })
                    .id()
            })
            .unwrap();

        app.update();

        let num_children = query_len!(app, (With<NodeStyleSheet>, With<Child>));
        assert_eq!(num_children, 2);

        app.world_mut()
            .run_system_once(
                move |mut commands: Commands, children_query: Query<&Children>| {
                    for child in children_query.iter_leaves(root_entity_id) {
                        commands
                            .entity(child)
                            .with_child(GrandChild)
                            .with_child(GrandChild);
                    }
                },
            )
            .unwrap();

        app.update();

        let num_grand_children = query_len!(app, (With<NodeStyleSheet>, With<GrandChild>));
        assert_eq!(num_grand_children, 4);

        assert_world_snapshot!(app, (Child, GrandChild, NodeStyleData));
    }

    #[test]
    fn marks_descendants_for_recalculation() {
        let mut app = app();

        let app = &mut app;

        app.add_systems(Startup, |mut commands: Commands| {
            commands
                .spawn((Name::new("Root"), ROOT))
                .with_children(|children| {
                    children
                        .spawn((FirstChild, Child))
                        .with_children(|children| {
                            children.spawn(GrandChild);
                            children.spawn(GrandChild);
                            children.spawn(GrandChild);
                        });
                    children.spawn(Child);
                    children.spawn(Child).with_children(|children| {
                        children.spawn(GrandChild);
                        children.spawn(GrandChild);
                        children.spawn(GrandChild);
                    });
                });
        });

        fn num_nodes_for_recalculation(app: &mut App) -> usize {
            let markers = query!(app, &NodeStyleMarker);
            markers.iter().filter(|c| c.needs_recalculation()).count()
        }

        fn clear_marked_for_recalculation(app: &mut App) {
            app.world_mut()
                .run_system_once(|mut query: Query<&mut NodeStyleMarker>| {
                    for mut computed_style in &mut query {
                        computed_style.clear_marker();
                    }
                })
                .unwrap();
        }

        // First run everything is marked for recalculation
        {
            app.update();
            assert_eq!(num_nodes_for_recalculation(app), 10);
            // We need to manually clear everything
            clear_marked_for_recalculation(app);
        }

        // With not updates, the number of nodes for recalculation should be still zero
        {
            app.update();
            assert_eq!(num_nodes_for_recalculation(app), 0);
        }

        {
            app.world_mut()
                .run_system_once(|mut query: Query<&mut NodeStyleMarker, Without<ChildOf>>| {
                    query.single_mut().mark_for_recalculation();
                })
                .unwrap();

            app.update();

            // Only root should be marked for recalculation again, since the flags says no recalculation is needed
            assert_eq!(num_nodes_for_recalculation(app), 1);

            clear_marked_for_recalculation(app);
        }

        // Only elements marked with Child should be marked
        {
            app.world_mut()
                .run_system_once(
                    |mut query: Query<(&NodeStyleData, &mut NodeStyleMarker), With<FirstChild>>| {
                        let (data, mut marker) = query.single_mut();
                        marker.mark_for_recalculation();
                        data.recalculation_flags.fetch_or(
                            RecalculateOnChangeFlags::RECALCULATE_SIBLINGS.bits(),
                            atomic::Ordering::Relaxed,
                        );
                    },
                )
                .unwrap();

            app.update();

            assert_eq!(num_nodes_for_recalculation(app), 3);

            clear_marked_for_recalculation(app);
        }

        // All elements should be marked
        {
            app.world_mut()
                .run_system_once(
                    |mut query: Query<(&NodeStyleData, &mut NodeStyleMarker), Without<ChildOf>>| {
                        let (data, mut marker) = query.single_mut();
                        marker.mark_for_recalculation();
                        data.recalculation_flags.fetch_or(
                            RecalculateOnChangeFlags::RECALCULATE_DESCENDANTS.bits(),
                            atomic::Ordering::Relaxed,
                        );
                    },
                )
                .unwrap();

            app.update();

            assert_eq!(num_nodes_for_recalculation(app), 10);

            clear_marked_for_recalculation(app);
        }
    }
}
