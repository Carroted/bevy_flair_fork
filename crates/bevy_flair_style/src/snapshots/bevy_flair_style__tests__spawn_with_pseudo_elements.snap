---
source: crates/bevy_flair_style/src/lib.rs
expression: scene_serializer
---
Scene(
  resources: {},
  entities: {
    4294967296: Entity(
      components: {
        "bevy_ecs::hierarchy::Children": Children([
          4294967297,
          4294967298,
          4294967299,
        ]),
        "bevy_ecs::name::Name": "Root",
        "bevy_flair_style::components::NodeStyleSheet": StyleSheet(Weak(Uuid(
          uuid: "fe981062-17ce-46e4-999a-5a61ea8fe722",
        ))),
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: None,
          previous_sibling: None,
        ),
      },
    ),
    4294967297: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967296),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: Some(4294967298),
          previous_sibling: None,
        ),
        "bevy_flair_style::tests::Child": Child(),
      },
    ),
    4294967298: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967296),
        "bevy_ecs::hierarchy::Children": Children([
          4294967300,
          4294967302,
          4294967301,
        ]),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: Some(4294967299),
          previous_sibling: Some(4294967297),
        ),
        "bevy_flair_style::tests::Child": Child(),
      },
    ),
    4294967299: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967296),
        "bevy_ecs::hierarchy::Children": Children([
          4294967303,
          4294967304,
        ]),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: None,
          previous_sibling: Some(4294967298),
        ),
        "bevy_flair_style::tests::Child": Child(),
      },
    ),
    4294967300: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967298),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::PseudoElement": Before,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: Some(4294967302),
          previous_sibling: None,
        ),
      },
    ),
    4294967301: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967298),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::PseudoElement": After,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: None,
          previous_sibling: Some(4294967302),
        ),
      },
    ),
    4294967302: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967298),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: Some(4294967301),
          previous_sibling: Some(4294967300),
        ),
      },
    ),
    4294967303: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967299),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::PseudoElement": Before,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: Some(4294967304),
          previous_sibling: None,
        ),
      },
    ),
    4294967304: Entity(
      components: {
        "bevy_ecs::hierarchy::ChildOf": ChildOf(4294967299),
        "bevy_flair_style::components::NodeStyleSheet": Inherited,
        "bevy_flair_style::components::PseudoElement": After,
        "bevy_flair_style::components::Siblings": Siblings(
          next_sibling: None,
          previous_sibling: Some(4294967303),
        ),
      },
    ),
  },
)
